glitch/parsers/docker_parser.py

```diff
@@ -0,0 +1,294 @@
+        files = [os.path.join(path, f) for f in os.listdir(path)]
+        dockerfiles = [f for f in files if os.path.isfile(f) and "Dockerfile" in os.path.basename(f)]
+        modules = [f for f in files if os.path.isdir(f) and DockerParser._contains_dockerfiles(f)]
+
+        blocks = [self.parse_file(f, UnitBlockType.script) for f in dockerfiles]
+        modules = [self.parse_module(f) for f in modules]
+        return blocks, modules
+
+    @staticmethod
+    def _contains_dockerfiles(path: str) -> bool:
+        if not os.path.exists(path):
+            return False
+        if not os.path.isdir(path):
+            return "Dockerfile" in os.path.basename(path)
+        for f in os.listdir(path):
+            contains_dockerfiles = DockerParser._contains_dockerfiles(os.path.join(path, f))
+            if contains_dockerfiles:
+                return True
+        return False
+
+    @staticmethod
+    def __parse_stage(name: str, path: str, unit_type: UnitBlockType, structure: list[DFPStructure]) -> UnitBlock:
+        u = UnitBlock(name, unit_type)
+        u.path = path
+        for s in structure:
+            DockerParser.__parse_instruction(s, u)
+
+        DockerParser.__merge_download_commands(u)
+        return u
+
+    @staticmethod
+    def __parse_instruction(element: DFPStructure, unit_block: UnitBlock):
+        instruction = element.instruction
+        if instruction in ['ENV', 'USER', 'ARG', 'LABEL']:
+            unit_block.add_variable(DockerParser.__create_variable_block(element))
+        elif instruction == 'FROM':
+            attr = Attribute("image", element.value.split(" ")[0], False)
+            attr.line = element.startline + 1
+            unit_block.add_attribute(attr)
+        elif instruction == 'COMMENT':
+            c = Comment(element.value)
+            c.line = element.startline + 1
+            unit_block.add_comment(c)
+        elif instruction in ['RUN', 'CMD', 'ENTRYPOINT']:
+            c_parser = CommandParser(element.value)
+            aus = c_parser.parse_command()
+            for attr in aus:
+                attr.line = element.startline
+            unit_block.atomic_units += aus
+        elif instruction == 'ONBUILD':
+            dfp = DockerfileParser()
+            dfp.content = element.value
+            element = DFPStructure(**dfp.structure[0])
+            DockerParser.__parse_instruction(element, unit_block)
+        elif instruction == 'COPY':
+            attr = AtomicUnit("", "copy")
+            src, dest = [v for v in element.value.split(" ") if not v.startswith("--")]
+            attr.add_attribute(Attribute('src', src, False))
+            attr.add_attribute(Attribute('dest', dest, False))
+            unit_block.add_atomic_unit(attr)
+        elif instruction in ['ADD', 'VOLUME', 'WORKDIR']:
+            pass
+        elif instruction in ['STOPSIGNAL', 'HEALTHCHECK', 'SHELL']:
+            pass
+        elif instruction == 'EXPOSE':
+            pass
```

Comment:
The 3 commands in the middle elif are rare and don't seem to add smells but I'll have to check them more since I didn't know of their existence.
The ADD adds a folder from the host machine to the docker image, VOLUME sets a shared volume between the host and container and the WORKDIR is basically a "cd" command with "mkdir".
The EXPOSE command works almost as documentation as referred in the docs https://docs.docker.com/engine/reference/builder/#expose

Due to these I skiped for know their implementation to focus on the smells with the intention of coming back to them

Reply:
Alright, then maybe you should add a TODO to these. The EXPOSE command might be interesting even as documentation. For instance, there may be a security issue with a certain port being opened. I do not know if this is true, but it is relevant information.